File: SkipList.java
import java.util.*;
import java.io.*;

class Node<T extends Comparable<T>> {

	T value;
	ArrayList<Node<T>> nextPointers;
	int height;

	Node(int height) {								// Constructor for Node
		this.nextPointers = new ArrayList<>(height);
		this.height = height;
		for(int i = 0; i < height; i++) {
			this.nextPointers.add(null);
		}
	}

	Node(T data, int height) {						// Another construct for node that also takes data
		this.nextPointers = new ArrayList<>();
		this.value = data;
		this.height = height;
		for(int i = 0; i < height; i++) {
			this.nextPointers.add(null);
		}
	}

	public T value() {			// Returns value of the node
		return this.value;
	}

	public int height() {		// Returns height of the node
		return this.height;
	}

	public Node<T> next(int level) {				// Returns the reference to the next node in the specified level
		if(level < 0 || level > this.height - 1) {
			return null;
		}

		return this.nextPointers.get(level);
	}

	// SUGGESTED METHODS //

	public void setNext(int level, Node<T> node) {	// Set the next reference at a specified level to point to a node
		this.nextPointers.set(level, node);
	}

	public void grow() {				// Grows the node by height 1
		this.height++;
		this.nextPointers.add(null);
	}

	public void maybeGrow() {			// Grows the node with a 50 percent probability
		double prob = 0.5;
		if(Math.random() < prob) {
			this.height++;
			this.nextPointers.add(null);
		}
	}

	public void trim(int height) { 		// Trims the node until it reaches a specified height
		if(this.value != null && this.height == 1) {
			return;
		}
		if(height >= this.height) {
			return;
		}
		for(int i = this.height-1; i >= height; i--) {	// Removes the top pointers until the height reaches the desired height
			this.nextPointers.remove(i);
			this.height--;
		}
		return;
	}
}

public class SkipList<T extends Comparable<T>> {

	Node<T> head;
	int height;
	int size;

	SkipList() { 					// Constructor to create a skiplist
		this.head = new Node<>(1);
		this.height = 1;
		this.size = 0;
	}

	SkipList(int height) {				// Another constructor to create a skiplist of a specific height
		this.head = new Node<>(height);
		this.height = height;
		this.size = 0;
	}

	public int size() {		// Returns the size of the skiplist
		return this.size; 
	}

	public int height() {	// Returns the height of the skiplist
		return this.height;
	}

	public Node<T> head() {	// Returns the reference to the head of the skiplist
		return this.head; 
	}

	public void insert(T data) {	// Inserts a value to the skiplist, in sorted order, with an optimal height
		Node<T> curr;
		int newsize = this.size+1;

		if(this.size == 0) {
			newsize++;
		}

		int log = (int)Math.ceil(lg(newsize));		
		int maxHeight = Math.max(log, this.height);			// Find the max suitable height for the node to be inserted in the list
		int randHeight = generateRandomHeight(maxHeight);	// Randomize a height that is withing constraints of a suitable height
		int newheight = this.height+1;				

		if(log > this.height) {								// Check if inserting the new node would require growing the skip list.
			growSkipList();
		}

		insert(data, randHeight);							// Use the insert method with the random generated height
		return;
	}

	public void insert(T data, int height) {			// Insert a node with a value into the list in O(log(n)) time

		Node<T> curr = this.head;
		Node<T> temp = new Node<T>(data, height);
		ArrayList<Node<T>> path = new ArrayList<>();	// ArrayList to store the traversal of the list, to keep track 
														// of references to be updated
		this.size++;

		for(int i = 0; i < this.height; i++) {			// Initializes the path ArrayList to store null references first
			path.add(null);
		}

		for(int i = this.height - 1; i >= 0; i--) {		// BST-style traversal of the skiplist
			if(curr.next(i) == null) {
				path.set(i, curr);
			}
			else if(data.compareTo(curr.next(i).value()) < 0) {
				path.set(i, curr);
			}
			else if(data.equals(curr.next(i).value())) {
				path.set(i, curr);
			}
			else {
				curr = curr.next(i);
				i++;
			}
		}

		for(int i = temp.height()-1; i >= 0; i--) {			// Updates references to point to the new node, and for the new node
				temp.setNext(i, path.get(i).next(i));		// to point to its next node
				path.get(i).setNext(i, temp);
		}
	}

	public void delete(T data) {							// Deletes a node (if exists) from the list in O(log(n)) time

		Node<T> curr = this.head;
		Node<T> want = null;								// This will be used to check if our node-to-be-deleted exists in the list
		ArrayList<Node<T>> path = new ArrayList<>();
		Node<T> temp = new Node<T>(data, height);

		for(int i = 0; i < this.height; i++) {				// Initialization for the path ArrayList
			path.add(null);
		}

		for(int i = this.height - 1; i >= 0; i--) {			// BST-style traversal of the skiplist
			if(curr.next(i) == null) {
				path.set(i, curr);
			}
			else if(data.compareTo(curr.next(i).value()) < 0) {
				path.set(i, curr);
			}
			else if(data.equals(curr.next(i).value())) {	// If we find our node-to-be-deleted, we store it
				path.set(i, curr);
				want = curr.next(i);
			}
			else {
				curr = curr.next(i);
				i++;
			}
		}

		if(want == null) {	// If we couldn't find our node-to-be-deleted, just return since there isn't anything to delete
			return;
		}

		for(int i = want.height()-1; i >= 0; i--) {		// Update references so we don't lose the references that our deleted node was pointing to.
			path.get(i).setNext(i, want.next(i));		// *Poof* we lost the references to our deleted node, so the Garbage Collector will get rid of it for us.
		}

		this.size--;	// Decrement the size of the skip list

		int newsize = this.size;
		int log = (int)Math.ceil(lg(newsize));

		if(newsize == 1 || newsize == 0) {
			log = 1;
		}

		while(log < this.height) {		// If our height is sub-optimal, we trim the list until the height becomes optimal
			trimSkipList();
		}
		return;
	}

	public boolean contains(T data) {			// Checks if the list contains a particular value in O(log(n)) time

		Node<T> curr = this.head;		

		for(int i = this.height - 1; i >= 0; i--) {		// BST-style traversal of the list
			if(curr.next(i) == null) {
			}
			else if(data.compareTo(curr.next(i).value()) < 0) {
			}
			else if(data.equals(curr.next(i).value())) { 
				return true;		// If we find our node, return true
			}	
			else {
				curr = curr.next(i);
				i++;
			}
		}
		return false; 	// If we couldn't find it, return false
	}

	public Node<T> get(T data) {		// Returns a reference to a specified node (if exists) in O(log(n)) time

		Node<T> curr = this.head;
		Node<T> want = null;

		for(int i = this.height - 1; i >= 0; i--) {				// BST-style traversal
			if(curr.next(i) == null) {
			}
			else if(data.compareTo(curr.next(i).value()) < 0) {
			}
			else if(data.equals(curr.next(i).value())) {
				want = curr.next(i);
				return want;		// If we find our node, return the reference to it
			}
			else {
				curr = curr.next(i);
				i++;
			}
		}

		return want; 	// If at this point we couldn't find our node, return null ('want' was initialized to null)
	}
	
	private static int generateRandomHeight(int maxHeight) {	// Generates a random height (within constraints) for a node.

		int level = 1;
		double prob = 0.5;

		while(Math.random() < prob) {	// Pretty much flipping a coin until you get an unwanted side.
			level++;
			if(level > maxHeight) {
				level = maxHeight;
				break;
			}
		}
		return level;
	}

	private void growSkipList() {	// Grows the skiplist by height 1.

		Node<T> curr = this.head().next(this.height);

		if(curr == null) {			// If there is nothing else at max height except the head, just grow the head.
			this.head.grow();
			this.height++;
			return;
		}

		do {						// Grow the top level nodes of the list by height 1
			curr.maybeGrow();
			curr = curr.next(this.height);
		} while(curr.next(this.height) != null);

		this.head.grow();			// Then grow the head
		this.height++;
	}
	

	private void trimSkipList() {	// Trims the skip list by height 1.

		Node<T> curr = this.head().next(this.height-1);
		Node<T> temp;

		if(curr == null) {					// If there is nothing at max height, just trim the head.
			this.head.trim(this.height-1);
			this.height--;
			return;
		}

		for(;;) {								// Trims all of the max-height nodes of the list by height 1
			temp = curr.next(this.height-1);
			curr.trim(this.height-1);
			if(temp == null) {
				break;
			}
			curr = temp;
		}

		this.head.trim(this.height-1);		// Then trim the head.
		this.height--;
	}

	public void print() {
        int curr_level = this.head.height();
        for(int i = curr_level - 1; i >= 0; i--) {
            Node<T> curr = this.head;
            while(curr != null) {
                System.out.print(((curr.value() == null) ? "null" : curr.value()) + " -> ");
                curr = curr.next(i);
            }
            System.out.println();
        }
        System.out.println("----");
    }

	private double lg(int a) {			// A base-2 logarithm function
		return logb(a, 2);
	} 

	private double logb(double a, double b) {	// A logartihm base converter using a property of logarithms
		return Math.log(a) / Math.log(b);
	}

	public static void main(String[] args) {

		SkipList<Integer> s = new SkipList<>();

		for(int i = 0; i < 100000; i++) {
			s.insert(i);
		}

		s.print();

		System.out.println(s.height());

	}
}

File: skiplist.cpp
#include <bits/stdc++.h>
#include <cmath>

using namespace std;

struct Node {

public:

	int priority;
	int value;
	int height;
	bool marked;

	vector<Node*> nextPointers;

	Node(int height) {
		this->height = height;
		for(int i = 0; i < height; i++) {
			(this->nextPointers).push_back(nullptr);
		}
	}

	Node(int data, int priority, int height) {
		this->value = data;
		this->height = height;
		this->priority = priority;
		for(int i = 0; i < height; i++) {
			(this->nextPointers).push_back(nullptr);
		}
	}

	int getValue() {
		return this->value;
	}

	int getPriority() {
		return this->priority;
	}

	int getHeight() {
		return this->height;
	}

	Node* getNext(int level) {
		if(level < 0 || level > this->height - 1) {
			return nullptr;
		}
		return this->nextPointers[level];
	}

	void setNext(int level, Node* node) {
		this->nextPointers[level] = node;
	}

	void grow() {
		(this->height)++;
		(this->nextPointers).push_back(nullptr);
	}

	void maybeGrow() {

		double prob = 50;
		int chance = rand() % 100 + 1;
		if(chance < prob) {
			(this->height)++;
			(this->nextPointers).push_back(nullptr);
		}
	}

	void trim(int height) {
		if(this->value != 0 && this->height == 1) {
			return;
		}
		if(height >= this->height) {
			return;
		}
		for(int i = (this->height) - 1; i >= height; i--) {
			(this->nextPointers).erase(this->nextPointers.begin() + i);
			this->height--;
		}
		return;
	}
};

struct SkipList {

public:

	Node* head;
	int height; 
	int size; 

	SkipList(void) {
		this->head = new Node(1);
		this->height = 1;
		this->size = 0;
	}

	SkipList(int height) {
		this->head = new Node(height);
		this->height = height;
		this->size = 0;
	}

	int getSize() {
		return this->size;
	}

	int getHeight() {
		return this->height;
	}

	Node* getHead() {
		return this->head;
	}

	int generateRandomHeight(int maxHeight) {

		int level = 1;
		double prob = 50.0;

		while((rand() % 100 + 1) < prob) {
			level++;
			if(level > maxHeight) {
				level = maxHeight;
				break;
			}
		}

		return level;
	}

	void growSkipList() {
		Node* curr = (this->getHead())->getNext(this->height);

		if(curr == NULL) {
			(this->head)->grow();
			this->height++;
			return;
		}

		do {
			curr->maybeGrow();
			curr = curr->getNext(this->height);
		} while(curr->getNext(this->height) != NULL);

		(this->head)->grow();
		this->height++;
	}

	void trimSkipList() {
		Node *curr = (this->getHead())->getNext(this->getHeight() - 1);
		Node* temp;

		if(curr == NULL) {
			this->getHead()->trim(this->getHeight() - 1);
			this->height--;
			return;
		}

		for(;;) {
			temp = curr->getNext(this->getHeight() - 1);
			curr->trim(this->getHeight() - 1);
			if(temp == NULL) {
				break;
			}
			curr = temp;
		}

		this->getHead()->trim(this->getHeight() - 1);
		this->height--;
	}

	void insert(int data, int priority, int height) {
		Node* curr = this->head;
		Node* temp = new Node(data, priority, height);
		vector<Node*> path;

		this->size++;

		for(int i = 0; i < this->height; i++) {
			path.push_back(nullptr);
		}

		for(int i = this->height - 1; i >= 0; i--) {
			if(curr->getNext(i) == nullptr) {
				path[i] = curr;
			}
			else if(priority < (curr->getNext(i))->getPriority()) {
				path[i] = curr;
			}
			else if(priority == (curr->getNext(i)->getPriority())) {
				path[i] = curr;
			}
			else {
				curr = curr->getNext(i);
				i++;
			}
		}

		for(int i = temp->getHeight() - 1; i >= 0; i--) {
			temp->setNext(i, path[i]->getNext(i));
			path[i]->setNext(i, temp);
		}
	}

	void insert(int data, int priority) {
		Node* curr;
		int newsize = this->size + 1;

		if(this->size == 0) {
			newsize++;
		}

		int log = (int)ceil((log2 (double(newsize))));
		int maxHeight = max(log, this->height);
		int randHeight = generateRandomHeight(maxHeight);
		int newheight = this->height+1;

		if(log > this->height) {
			growSkipList();
		}

		insert(data, priority, randHeight);
		return;

	}

	void sldelete(int data, int priority) {
		Node* curr = this->head;
		Node* want = NULL;
		vector<Node*> path;

		for(int i = 0; i < this->height; i++) {
			path.push_back(nullptr);
		}

		for(int i = this->height - 1; i >= 0; i--) {
			if(curr->getNext(i) == NULL) {
				path[i] = curr;
			}
			else if(data < (curr->getNext(i))->getValue()) {
				path[i] = curr;
			}
			else if(data == curr->getNext(i)->getValue()) {
				path[i] = curr;
				want = curr->getNext(i);
			}
			else {
				curr = curr->getNext(i);
				i++;
			}
		}

		if(want == NULL) {
			return;
		}

		for(int i = want->getHeight() - 1; i >= 0; i--) {
			path[i]->setNext(i, want->getNext(i));
		}

		this->size--;

		int newsize = this->size;
		int log = (int)ceil((log2 (double(newsize))));

		if(newsize == 1 || newsize == 0) {
			log = 1;
		}

		while(log < this->height) {
			trimSkipList();
		}

		return;
	}

	// void sldelete(int data, int priority) {
	// 	Node* curr = this->head;
	// 	Node* want = NULL;
	// 	vector<Node*> path;

	// 	for(int i = 0; i < this->height; i++) {
	// 		path.push_back(nullptr);
	// 	}

	// 	for(int i = this->height - 1; i >= 0; i--) {
	// 		if(curr->getNext(i) == NULL) {
	// 			path[i] = curr;
	// 		}
	// 		else if(data < (curr->getNext(i))->getValue()) {
	// 			path[i] = curr;
	// 		}
	// 		else if(data == curr->getNext(i)->getValue()) {
	// 			path[i] = curr;
	// 			want = curr->getNext(i);
	// 		}
	// 		else {
	// 			curr = curr->getNext(i);
	// 			i++;
	// 		}
	// 	}

	// 	if(want == NULL) {
	// 		return;
	// 	}

	// 	for(int i = want->getHeight() - 1; i >= 0; i--) {
	// 		path[i]->setNext(i, want->getNext(i));
	// 	}

	// 	this->size--;

	// 	int newsize = this->size;
	// 	int log = (int)ceil((log2 (double(newsize))));

	// 	if(newsize == 1 || newsize == 0) {
	// 		log = 1;
	// 	}

	// 	while(log < this->height) {
	// 		trimSkipList();
	// 	}

	// 	return;
	// }

	void delete_min() {

		Node* curr = this->head;
		Node* want = NULL;
		vector<Node*> path;

		for(int i = 0; i < this->height; i++) {
			path.push_back(nullptr);
		}

		int top_priority = INT_MAX;

		for(int i = this->height - 1; i >= 0; i--) {
			path[i] = curr;
		}

		want = this->head->getNext(0);

		for(int i = want->getHeight() - 1; i >= 0; i--) {
			path[i]->setNext(i, want->getNext(i));
		}

		this->size--;

		int newsize = this->size;
		int log = (int)ceil((log2 (double(newsize))));

		if(newsize == 1 || newsize == 0) {
			log = 1;
		}

		while(log < this->height) {
			trimSkipList();
		}

		return;
	}


	bool contains(int data) {
		Node* curr = this->head;
		Node* want = NULL;

		for(int i = this->getHeight() - 1; i >= 0; i--) {
			if(curr->getNext(i) == NULL) {

			}
			else if(data < (curr->getNext(i))->getValue()) {
			}
			else if(data == (curr->getNext(i))->getValue()) {
				return true;
			}
			else {
				curr = curr->getNext(i);
				i++;
			}
		}
		return false;
	}

	Node* get(int data) {
		Node* curr = this->getHead();
		Node* want = NULL;

		for(int i = this->getHeight() - 1; i >= 0; i--) {
			if(curr->getNext(i) == NULL) {
			}
			else if(data < (curr->getNext(i))->getValue()) {
			}
			else if(data == (curr->getNext(i))->getValue()) {
				want = curr->getNext(i);
				return want;
			}
			else {
				curr = curr->getNext(i);
				i++;
			}
		}

		return want;
	}

	void print(int mode) {

        ofstream myfile;

        if(mode == 0) {
        	myfile.open("outputCPP.txt");
        }
        else if(mode == 1) {
        	myfile.open("outputCPP.txt", ios_base::app);
        }
        

        int curr_level = head->height;

        for(int i = curr_level - 1; i >= 0; i--) {
            Node *curr = head->getNext(i);

            while(curr != NULL) {
                ((curr == NULL) ? myfile << "NULL" : myfile << "[" << curr->getPriority() << ", " << curr->getValue()) << "]" << " -> ";
                curr = curr->getNext(i);
            }
            myfile << endl;
        }
        myfile << "----" << endl;
       	myfile << "height: " << this->getHeight() << endl;
        myfile.close();
    }

    void printHeight() {
    	cout << "height: " << this->getHeight() << endl;
    }

};

int main() {

	srand(time(NULL));

	SkipList *s = new SkipList();

	clock_t begin, end;

	begin = clock();

	for(int i = 1; i <= 1000000; i++) {
		s->insert(123124, i);
		if(s->getHeight() >= 14) {
			// cout << "height: " << s->getHeight() << endl;
			while(true) {
				s->delete_min();
				if(s->height == 12) {
					break;
				}
			}
		}
	}

	end = clock();

	double elapsed_secs = (double(end - begin) / CLOCKS_PER_SEC);

	// int height1 = s->getHeight();

	// s->print(0);

	// for(int i = 0; i <= 800000; i++) {
	// 	s->delete_min();
	// }

	// int height2 = s->getHeight();

	cout << "delete: " << elapsed_secs * 1000 << " ms" << endl;

	// cout << "prev height: " << height1 << " new height: " << height2 << endl;

	// s->print(1);

}